State management service - Monkfish (GSI-898) (#119)
ghga-de/epic-docs: 57-monkfish/technical_specification.md

    on the specified collection.
- `PUT /docs/{db-name}/{collection-name}`
  - *Upserts the document(s) provided in the request body in the specified collection.*
  - Request body:
    - Single document to upsert, or array of documents for batch upsertion
  - Authorization header: API Key (set in configuration)
  - Response body: Empty
  - Response status:
    - `204 No Content`: Document(s) successfully upserted
    - `401 Unauthorized`: auth error (not authenticated)
    - `403 Forbidden`: Authenticated, but config prevents operation
    on the specified collection.
- `DELETE /docs/{db-name}/{collection-name}`
  - *Deletes all or some documents in the collection.*
  - Query string (optional):
    - Filter parameters to refine deletion, e.g. `user_id=123&name=keith`
  - Authorization header: API Key (set in configuration)
  - Response body: Empty
  - Response status: 
    - `204 No Content`: Document(s) did not exist or were successfully deleted
    - `401 Unauthorized`: Auth error (not authenticated)
    - `403 Forbidden`: Authenticated, but config prevents operation
    on the specified collection.

## Human Resource/Time Estimation:

> Number of sprints required: 2

Author: For extensions also

## Scope
### Outline:
The [Archive Test Bed](https://github.com/ghga-de/archive-test-bed) needs a way to
consistently perform arbitrary CRUD operations
in MongoDB that are outside the scope of the component services, but also unwieldy
to manage in piecemeal code adaptations. A dedicated service can offer a simple
solution to this problem and allow test bed services and processes to programmatically
seed, reset, and modify databases through a single API.

### Included/Required:
The implementation of the SMS should include a token-secured RESTful API that interacts
with MongoDB.

### Not included:
Object schema validation would be complex to add because the models can come from a
variety of sources (`ghga-event-schemas`, service-specific models, etc.), and keeping
this service in sync with those definitions, either through config or code, would
incur a cost outweighing the benefits provided by the service.

## API Definitions:

The following REST endpoints will be created:

- `GET /{db-name}/{collection-name}`
  - *Returns all resources in the collection.*
  - Query string (optional):
    - Filter parameters to refine results, e.g. `user_id=123`

>   - Auth header: internal token

Reviewer: ```suggestion
  - Authorization header: Apikey (set in configuration)
```

  - Response body: list of resources in the collection matching the specified criteria
  - Response status: 
    - `200 OK`: Request successfully processed
    - `401 Unauthorized`: auth error (not authenticated)

> - `PUT /{db-name}/{collection-name}/{id}`

Reviewer: ```suggestion
- `PUT /{db-name}/{collection-name}`
```

Reviewer: Similar to the other services, should not have the id in the path. Instead, the id is taken from the document provided in the request body. If no ID exists there, there should be an error (the user is supposed to generate one for an insertion).

>   - *Upserts the resource with the given ID using the payload provided in the request body.*

Reviewer: ```suggestion
  - *Upserts the document(s) provided in the request body in the specified collection.*
```

  - Request body:

>     - Resource to upsert

Reviewer: ```suggestion
    - Single document to upsert, or array of documents for batch upsertion
```

  - Auth header: internal token
  - Response body:

>     - `id`: string (the ID of the upserted resource)

Reviewer: Can be empty, since the user provides the ID already.

  - Response status:

>     - `201 Created`: resource successfully upserted

Reviewer: Maybe better just `204 No Content` since it could be an update without creation of any new resources.

    - `401 Unauthorized`: auth error (not authenticated)
- `DELETE /{db-name}/{collection-name}`

>   - *Deletes the specified collection.*

Reviewer: ```suggestion
  - *Deletes all or some documents in the collection.*
```

  - Query string (optional):

>     - Filter parameters to refine deletion, e.g. `user_id=123`

Reviewer: see above

  - Auth header: internal token

>   - Response body: list of deleted IDs

Reviewer: Maybe just empty body, and status code 204.

consistently perform arbitrary CRUD operations
in MongoDB that are outside the scope of the component services, but also unwieldy
to manage in piecemeal code adaptations. Namely, the needs are:
- setting a predefined state before running the tests (empty or prepopulate databases)
- examining the state after or in between tests (whitebox testing)

Right now, there's not a clean way to achieve either. 
A dedicated service can offer a simple solution to these problems and allow test bed 
processes to programmatically seed, reset, modify and examine databases through a single API.

This service should **never** be deployed to production. It is *only* intended for use
in the Testing and Staging environments, where there is no access to real data.  
Despite this, there should be a way to restrict which databases and collections can be
accessed with this service through configuration, and a simple API key (set in config)
that can be used to authenticate requests.

### Included/Required:
The implementation of the SMS should include a token-secured RESTful API that interacts
with MongoDB. There should be configuration to control access to databases and collections,

> as well as an API Key for authentication.

Reviewer: You can already specify here the granularity of the access control config.

Suggestion: the setting should be a list of tuples containing (db-name, collection-name, permissions). permissions can we a string with the letters "crud". A value of none indicates no restriction. The first matching rule will be applied. A value of None can be used as a wildcard. If no rule matches, access will be denied.

E.g. `[('auth', 'users', 'c'), ('store', 'orders', 'ud'), (None, None, 'r')]` to allow creating users in db auth, updating or deleting orders in db store, and read from any other collection in any database. Or `[(None, None, None)]` to allow anything on any database.

Reviewer: Or something similar, if you have a better idea.

### Not included:
Object schema validation would be complex to add because the models can come from a
variety of sources (`ghga-event-schemas`, service-specific models, etc.), and keeping
this service in sync with those definitions, either through config or code, would
incur a cost outweighing the benefits provided by the service.

## API Definitions:

The following REST endpoints will be created:

- `GET /{db-name}/{collection-name}`
  - *Returns all or some documents from the collection.*
  - Query string (optional):
    - Filter parameters to refine results, e.g. `user_id=123&role=supervisor` 
  - Authorization header: API Key (set in configuration)
  - Response body: list of resources in the collection matching the specified criteria
  - Response status: 

>     - `200 OK`: Request successfully processed

Reviewer: ```suggestion
    - `200 OK`: Request successfully processed, results in response body
```

    - `401 Unauthorized`: auth error (not authenticated)
- `PUT /{db-name}/{collection-name}`
  - *Upserts the document(s) provided in the request body in the specified collection.*
  - Request body:
    - Single document to upsert, or array of documents for batch upsertion
  - Authorization header: API Key (set in configuration)
  - Response body: Empty
  - Response status:

>     - `204 No Content`: Document successfully upserted

Reviewer: ```suggestion
    - `204 No Content`: Document(s) successfully upserted
```

    - `401 Unauthorized`: auth error (not authenticated)
- `DELETE /{db-name}/{collection-name}`
  - *Deletes all or some documents in the collection.*
  - Query string (optional):
    - Filter parameters to refine deletion, e.g. `user_id=123&name=keith`
  - Authorization header: API Key (set in configuration)
  - Response body: Empty
  - Response status: 

>     - `204 No Content`: Document(s) successfully deleted

Reviewer: ```suggestion
    - `204 No Content`: Document(s) did not exist or were successfully deleted
```

processes to programmatically seed, reset, modify and examine databases through a single API.

This service should **never** be deployed to production. It is *only* intended for use
in the Testing and Staging environments, where there is no access to real data.  
Despite this, there should be a way to restrict which databases and collections can be
accessed with this service through configuration, and a simple API key (set in config)
that can be used to authenticate requests.

### Included/Required:
The implementation of the SMS should include a token-secured RESTful API that interacts
with MongoDB. There should be configuration to control access to databases and collections,
as well as an API Key for authentication.

Access control for database/collection resources should occur on a whitelist basis, with
further specification for the individual permissions granted at the collection level
for create, read, update, and delete.

Since database names in the testing environment feature a prefix for isolation, the
config should have a `db_prefix` value. This will be prefixed to all supplied database
names, meaning the full name doesn't have to be supplied by the user in configuration,

> making the configuration more readable.

Reviewer: Not only for the config, the `db_prefix` should also be silently added to all db names specified as path parameters.

`None` or `*` should be used as a wildcard equivalent to 'allow all'.  
For example, using the wildcard to specify permissions for a collection will allow all
CRUD operations on the collection. Using the wildcard to specify tables in a database

> will enable all CRUD operations on all collections in that database.

Reviewer: Not necessarily. You could also specify a rule where the table name is a wildcard, but the permissions are restriced e.g. to only allow reading, but not writing.

If neither `db_prefix` nor any database/collections are listed, then no operations will

> be allowed.

Reviewer: Even if there is a `db_prefix`, if no datatabases are listed, still nothing should be allowed.

The only effect of the `db_prefix` should be that it will be silently added to all database operations. Thereby, the prefix does not need to be added to database names specified in the whitelist or as path parametes.

This also guarantees that only databases with that prefix can be access via the service. To allow full access to all of these database, you would still need to add a single wildcard rule allowing everything on all databases.

If no database/collections are listed, then no operations will be allowed, even if
`db_prefix` is specified.

## Apache Kafka, S3, Vault

Similar to the fixture state reset logic featured in `hexkit`, the utilities available
for Apache Kafka should enable deleting and publishing records (events) in a topic.  
The advantage over using Kafka UI is that the SMS centralized state management for other

> technologies and can be used programmatically.

Author: This is a good point, I'll add that.
